File: web-ui/src/types/graph.ts
import type { Node, Edge } from '@xyflow/react'

export type ActorType = 'supervisor' | 'source' | 'transform' | 'sink' | 'router'
export type ActorState = 'created' | 'starting' | 'running' | 'stopping' | 'stopped' | 'restarting' | 'failed'

export interface ActorMetrics {
  input_count: number
  output_count: number
  error_count: number
  throughput_per_sec: number
  state: ActorState
  last_updated: string
}

export interface GraphNodeData {
  actor_name: string
  actor_type: ActorType
  actor_config?: Record<string, unknown>
  parallelism?: number
  metrics?: ActorMetrics
  [key: string]: unknown  // Index signature for React Flow compatibility
}

export interface GraphPosition {
  x: number
  y: number
}

export interface GraphNode {
  id: string
  type: string
  label: string
  position: GraphPosition
  data: GraphNodeData
}

export interface GraphEdge {
  id: string
  source: string
  target: string
  label?: string
  animated?: boolean
}

export interface PipelineGraph {
  pipeline_id: string
  nodes: GraphNode[]
  edges: GraphEdge[]
  layout?: string
}

export interface GraphUpdateRequest {
  add_edges?: GraphEdge[]
  remove_edges?: string[]
  update_nodes?: Array<{
    id: string
    position: GraphPosition
  }>
}

// React Flow 호환 타입
export type ActorNode = Node<GraphNodeData>
export type PipelineEdge = Edge

// API 응답 타입
export interface ApiResponse<T> {
  success: boolean
  data?: T
  error?: string
}

export type ActorMetricsMap = Record<string, ActorMetrics>

File: shared/types/graph.go
package types

import "time"

// GraphNode React Flow 호환 노드 구조
type GraphNode struct {
	ID       string         `json:"id"`
	Type     string         `json:"type"` // source, transform, sink, router, supervisor
	Label    string         `json:"label"`
	Position GraphPosition  `json:"position"`
	Data     *GraphNodeData `json:"data"`
}

// GraphPosition 노드 위치 좌표
type GraphPosition struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

// GraphNodeData 노드 상세 데이터
type GraphNodeData struct {
	ActorName   string         `json:"actor_name"`
	ActorType   string         `json:"actor_type"`
	ActorConfig map[string]any `json:"actor_config,omitempty"`
	Parallelism int            `json:"parallelism,omitempty"`
	Metrics     *ActorMetrics  `json:"metrics,omitempty"`
}

// ActorMetrics Actor별 메트릭
type ActorMetrics struct {
	InputCount       int64      `json:"input_count"`
	OutputCount      int64      `json:"output_count"`
	ErrorCount       int64      `json:"error_count"`
	ThroughputPerSec float64    `json:"throughput_per_sec"`
	State            ActorState `json:"state"`
	LastUpdated      time.Time  `json:"last_updated"`
}

// GraphEdge 노드 간 연결
type GraphEdge struct {
	ID       string `json:"id"`
	Source   string `json:"source"`
	Target   string `json:"target"`
	Label    string `json:"label,omitempty"`
	Animated bool   `json:"animated,omitempty"`
}

// PipelineGraph 전체 파이프라인 그래프 구조
type PipelineGraph struct {
	PipelineID string      `json:"pipeline_id"`
	Nodes      []GraphNode `json:"nodes"`
	Edges      []GraphEdge `json:"edges"`
	Layout     string      `json:"layout,omitempty"` // "dagre" or "manual"
}

// GraphUpdateRequest 그래프 연결 업데이트 요청
type GraphUpdateRequest struct {
	AddEdges    []GraphEdge       `json:"add_edges,omitempty"`
	RemoveEdges []string          `json:"remove_edges,omitempty"`
	UpdateNodes []GraphNodeUpdate `json:"update_nodes,omitempty"`
}

// GraphNodeUpdate 노드 위치 업데이트
type GraphNodeUpdate struct {
	ID       string        `json:"id"`
	Position GraphPosition `json:"position"`
}

// ActorMetricsMap Actor별 메트릭 맵 (polling 응답용)
type ActorMetricsMap map[string]*ActorMetrics
